compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -2097,42 +2103,44 @@ where
             }
 
             // Arrays and slices.
-            ty::Array(element, _) | ty::Slice(element) => element,
-            ty::Str => tcx.types.u8,
+            ty::Array(element, _) | ty::Slice(element) => Ok(element),
+            ty::Str => Ok(tcx.types.u8),
 
             // Tuples, generators and closures.
-            ty::Closure(_, ref substs) => substs.as_closure().upvar_tys().nth(i).unwrap(),
+            ty::Closure(_, ref substs) => {
+                Self::ty_and_layout_kind(this, cx, i, substs.as_closure().tupled_upvars_ty())
```

Comment:
I see, so the motivation is so that we can redirect here? And in particular so that we can get the field of a type that is not `self.ty`? You could also just have this arm of the match use a `return`...

```rust
cx.layout_of(match foo {
    ...
    Closure => return ...,
    ...
})
```

though arguably it's harder to follow. I'm ok with a helper, just wanted to understand its motivation better.

Reply:
Yes, that was the idea. I have moved the helper function so it now inside of the `field` method.